src/bridged_logging.rs

```diff
@@ -0,0 +1,245 @@
+
+extern crate log as rs_log;
+
+use glib_sys;
+use translate::*;
+
+/// Enumeration of the possible formatting behaviours for a [`GlibLogger`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum GlibLoggerFormat {
+    /// A simple format, writing only the message on output.
+    Plain,
+    /// A simple format, writing file, line and message on output.
+    LineAndFile,
+    /// A logger using glib structured logging.
+    #[cfg(any(feature = "v2_56", feature = "dox"))]
+    Structured,
+}
+
+/// Enumeration of the possible domain handling behaviours for a [`GlibLogger`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum GlibLoggerDomain {
+    /// Logs will have no domain specified.
+    None,
+    /// Logs will use the `target` of the log crate as a domain; this allows
+    /// Rust code like `warn!(target: "my-domain", "...");` to log to the glib
+    /// logger using the specified domain.
+    CrateTarget,
+    /// Logs will use the crate path as the log domain.
+    CratePath,
+}
+
+/// An implementation of a `log` crate logger which logs over glib logging
+/// facilities.
+///
+/// Example:
+///
+/// ```no_run
+/// static glib_logger: glib::GlibLogger = glib::GlibLogger::new(
+///     glib::GlibLoggerFormat::Structured,
+///     glib::GlibLoggerDomain::CrateTarget,
+/// );
+///
+/// fn main() {
+///     log::set_logger(&glib_logger);
+///     log::set_max_level(log::LevelFilter::Debug);
+///
+///     info!("This line will get logged by glib");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct GlibLogger {
+    format: GlibLoggerFormat,
+    domain: GlibLoggerDomain,
+}
+
+impl GlibLogger {
+    /// Creates a new instance of `GlibLogger`.
+    ///
+    /// Example:
+    ///
+    /// ```no_run
+    /// static glib_logger: GlibLogger = glib::GlibLogger::new(
+    ///     glib::GlibLoggerFormat::Structured,
+    ///     glib::GlibLoggerDomain::CrateTarget,
+    /// );
+    ///
+    /// fn main() {
+    ///     log::set_logger(&glib_logger);
+    ///     log::set_max_level(log::LevelFilter::Debug);
+    ///
+    ///     info!("This line will get logged by glib");
+    /// }
+    /// ```
+    pub const fn new(format: GlibLoggerFormat, domain: GlibLoggerDomain) -> Self {
+        GlibLogger { format, domain }
+    }
+
+    fn level_to_glib(level: rs_log::Level) -> glib_sys::GLogLevelFlags {
+        match level {
+            // Errors are mapped to critical to avoid automatic termination
+            rs_log::Level::Error => glib_sys::G_LOG_LEVEL_CRITICAL,
+            rs_log::Level::Warn => glib_sys::G_LOG_LEVEL_WARNING,
+            rs_log::Level::Info => glib_sys::G_LOG_LEVEL_INFO,
+            rs_log::Level::Debug => glib_sys::G_LOG_LEVEL_DEBUG,
+            // There is no equivalent to trace level in glib
+            rs_log::Level::Trace => glib_sys::G_LOG_LEVEL_DEBUG,
+        }
+    }
+
+    fn write_log(domain: Option<&str>, level: rs_log::Level, message: &str) {
```

Comment:
Just wondering: isn't this function supposed to be public?

Reply:
No, it's an internal function called from the public one.